README.md
# artery

This is a template for running Express powered APIs on Azure App Service with:

- Blue-green zero-downtime deployments and all environments behind Azure AD auth
- Monitoring, alerts, availability tests and logging to analytics workspace
- OpenAPI middleware validating requests and routing to (TypeScript) handlers
- Azure Pipelines for updating Azure services and releasing Docker images to ACR
- A pipeline for importing/updating the API in existing Azure API Management(s)

Principles:

- OpenAPI driven development for mocking and enforcing req and res validations
- Generate both API tests and load tests automatically based on API definition
- Minimize number of environments, internal differences, maintenance and costs
- One command deploy PaaS with connectivity, monitoring and identity services
- Create Azure DevOps project and pipelines with IAC, requiring only Azure CLI

## 🔨 Development

Install development tools with [Homebrew](https://brew.sh/) on macOS and
Linux distros:

    brew bundle

If you cannot use Homebrew, install them manually.

Install Node.js version in `.nvmrc`:

    nvm install

Install Node.js packages:

    npm install

### Code

Build production `dist/`:

    npm run build

Start development server that watches for changes:

    npm run dev

### Test

Run functional tests on [Jest](https://jestjs.io/):

    npm test

Run API tests on [Newman](https://github.com/postmanlabs/newman):

    npm run test:api

Generate and run [k6](https://k6.io/) load tests and this API in Docker Compose:

    npm run k6

The Docker Compose stack includes
[Grafana dashboard](http://localhost:3000/d/k6/k6-load-testing-results)
for load test results:

![K6 load test results dashboard in Grafana](docs/grafana_k6.png)

### Observe

[Morgan](https://github.com/expressjs/morgan) is used as Express middleware for
logging HTTP requests to console.

See [localhost:8080/stats](http://localhost:8080) and login with username
and password `local` for real-time metrics by
[swagger-stats middleware](https://github.com/slanatech/swagger-stats).

![Real-time statistics for endpoints](docs/swagger_stats.png)

When [deployed to Azure](bicep/README.md) username and password are
set in the App Service app settings and generated from the branch name and
from the commit SHA by the [Azure DevOps pipeline](devops/README.md).

In Azure, Node.js metrics are streamed to Application Insights and logs are
sent to Log Analytics Workspace.

## 🧰 Practices

### Scan vulnerabilities

Have [Trivy](https://trivy.dev) present in your system to run vulnerability,
secret, misconfiguration and license scans automatically on commit.

You may run the scans manually in the git working copy:

    npm run sec

Or to write `results/` in various different formats:

    npm run sec:html
    npm run sec:junit
    npm run sec:sarif

### Use pre-commit checks

[Husky](https://typicode.github.io/husky/#/) is used for running the static
code analysis tools defined in `.husky/pre-commit`:

- [Trivy](https://trivy.dev) for various security scans
- [Prettier](https://prettier.io/) for formatting code based on `.prettierrc`
- [ESLint](https://eslint.org/) for linting and fixing code based on `.eslintrc.js`
- [Spectral](https://stoplight.io/open-source/spectral/) for linting OpenAPI
- [Portman](https://github.com/apideck-libraries/portman) for updating API tests
  based on `openapi.yaml`

Husky is installed in Node.js development dependencies (on `npm install`).

### Use IDE checks

Use VSCode to install the VSCode extensions defined in `.vscode/extensions.json`
so linting, static analysis and formatting runs continuously write-time,
before code even ends up in a commit.

Have [Semgrep](https://github.com/returntocorp/semgrep) and
[hadolint](https://github.com/hadolint/hadolint) present in your system
to benefit from the VSCode extensions.

#### Linting

You may run [ESLint](https://eslint.org/) manually for linting the codebase:

    npm run lint:fix

Or [Spectral](https://stoplight.io/open-source/spectral/) manually for linting
the OpenAPI files:

    npm run lint:spec

#### Static analysis

You may run [Semgrep](https://github.com/returntocorp/semgrep) manually for the
codebase:

    npm run sa

### Virtualize APIs

Use [Prism](https://github.com/stoplightio/prism) locally to fake the APIs
defined in `openapi.yaml` that are not yet implemented:

    npm run prism:mock

Fake data is generated dynamically according to the `x-faker` properties in the
OpenAPI definition. See [Faker.js](https://github.com/marak/Faker.js#api-methods)
for all different kinds of test data to generate.

To run Prism so that the already implemented endpoints are responded by Express:

    npm run prism:proxy

Prism server is launched at [localhost:4010](http://localhost:4010), returning
mock responses for the non-implemented endpoints according to
`example`/`examples` properties in the OpenAPI definition (`openapi.yaml`).

This applies for all the routes which do not have handler set by the OpenAPI
properties `operationId` and `x-eov-operation-handler`: The Express will
respond HTTP status 501 (Not Implemented) for routes that do not have handlers
implemented, causing Prism then to mock the response.

## ⚙️ CI/CD

The git workflow is [trunk based development](https://trunkbaseddevelopment.com/).

The branches map to pipeline stages and target Azure environments as following:

| Branch |   Stage    | CI/CD | CDP | Deployment trigger                       | Cluster |
| ------ | :--------: | ----- | :-: | ---------------------------------------- | :-----: |
| `*`    |     PR     | ✔️    | ✔️  | The tests in the feature branch passed   |   stg   |
| `main` |  staging   | ✔️    | ✔️  | A PR was merged and tests passed in main |   stg   |
| `main` |     rc     | ✔️    | ✔️  | Post-deployment gates in staging passed  |  prod   |
| `main` | production | ✔️    |     | A person runs the production pipeline    |  prod   |

Stages `rc` and `production` re-use the artifacts from `staging`.

The only task of the `production` pipeline is to roll out `rc` to `production`.

The infrastructure deployment pipelines are separated from the app deployment,
except for pull request review environments, the infrastructure is created in
the pipeline.

### Azure App Service

The stages are implemented as slots in `stg` and `prod` App Services.

Swapping the slots is used to roll out the new app version.

Slots can be swapped back to do a fast roll back (unless have DB migrations).

App Service can be configured to canaries and progressive delivery using slots.

### Azure DevOps

Azure DevOps pipelines are used for building Docker images and deploying the
images to environments in Azure.

To create the Azure DevOps project and the pipelines in your Azure DevOps
organisation, see
[creating Azure DevOps resources programmatically](devops/README.md).

For **public Azure DevOps projects, don't allow PRs to be deployed**
without a review from a team member as doing so might compromise security.

## 🏗️ Infrastructure

Docker images are used as the primary distribution mechanism, thus this APIs
npm package is configured as private in `package.json`.

The Docker images can be hosted in any Docker registry and the containers will
run in any system having Docker present.

Regardless, Azure Container Registry and Azure App Service are preferred here
as the container registry and as the PaaS.

### Docker images

Multi-staged `docker/` files build the images as following:

- The first stage runs all tests before transpiling TypeScript to JavaScript
- The second stage copies JS files and installs only the production dependencies
- The final run-time image knows nothing about packages and only has pure JS

To build the API image locally and run the production server as container:

    docker/build_and_test_image

Or to run the API, as well as k6 and Grafana with Docker Compose:

    docker compose up

By default, Alpine Linux based Docker images are built, but Debian Bullseye
(slim) Dockerfiles are also included in `docker/` if
[Alpine caveats](https://github.com/gliderlabs/docker-alpine/blob/master/docs/caveats.md) are present in your system.

Docker images are built by the CI/CD pipelines per environment and pushed to
the environment specific Azure Container Registry. When building the image,
`docker/build_and_test_image` and `docker/Dockerfile.alpine.prod` are used.

The only difference between non-prod and prod images is that non-prod will
install, configure (according to App Service's guidelines) and start the
SSH daemon which can help debugging Azure App Service in pull requests.

### Infrastructure-as-Code

Azure resources are designed to be updated by the _infra_ Azure DevOps pipeline
after initially having [created the Azure resources](bicep/README.md) for the
environments.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 1.0.0 (2022-10-15)


src/__tests__/index.test.ts
import { spawn, ChildProcess } from 'child_process'
import axios, { AxiosInstance } from 'axios'
import waitOn from 'wait-on'

jest.setTimeout(15000)

const serverPort = 8080

describe('HTTP endpoint tests', () => {
  let start: ChildProcess
  let client: AxiosInstance

  beforeAll(async () => {
    client = axios.create({
      baseURL: `http://localhost:${serverPort}`,
      validateStatus: () => true
    })
    const env = process.env
    env.SWAGGER_STATS_USERNAME = 'local'
    env.SWAGGER_STATS_PASSWORD = 'local'
    start = spawn('npm', ['start'], {
      cwd: __dirname,
      env,
      detached: true,
      stdio: 'inherit'
    })
    await waitOn({ resources: ['tcp:localhost:8080'] })
  })

  afterAll(() => {
    if (typeof start.pid !== 'undefined') {
      process.kill(-start.pid)
    }
  })

  test('GET to / responds 200', async () => {
    const res = await client.get('/')
    expect(res.status).toBe(200)
  })

  test('GET to /favicon.ico responds 200', async () => {
    const res = await client.get('/favicon.ico')
    expect(res.status).toBe(200)
  })

  test('GET to /stats responds 200', async () => {
    const res = await client.get('/stats')
    expect(res.status).toBe(200)
  })

  test('GET to non-implemented responds 501', async () => {
    const res = await client.get('/users')
    expect(res.status).toBe(501)
  })

  test('GET to non-existing responds 404', async () => {
    const res = await client.get('/not-found')
    expect(res.status).toBe(404)
  })
})


src/handlers.ts
import fs from 'fs'
import path from 'path'

import axios from 'axios'

// version
let version = 'dev'
try {
  version = fs.readFileSync(path.join(__dirname, '..', 'builtInfo.txt'), 'utf8')
} catch (err) {
  if (err.code !== 'ENOENT') throw err
}

// environment variables
const privateBackendUrl =
  process.env.PRIVATE_BACKEND_URL || 'https://duckduckgo.com'

// handlers
module.exports = {
  version: async (req: any, res: any) => {
    await res
      .status(200)
      .json({ version, serverTimeUTC: new Date().toISOString() })
  },
  spec: async (req: any, res: any) => {
    await res.status(200).sendFile(path.join(__dirname, '..', 'openapi.yaml'))
  },
  docs: async (req: any, res: any) => {
    await res.status(200).send(`
      <!doctype html>
        <html>
        <head>
          <meta charset="utf-8">
          <script type="module" src="https://unpkg.com/rapidoc/dist/rapidoc-min.js"></script>
        </head>
        <body>
          <rapi-doc
            spec-url = "/spec"
            theme = "dark"
            show-header = "false"
          ></rapi-doc>
        </body>
        </html>`)
  },
  dnsServers: async (req: any, res: any) => {
    await res.status(200).sendFile('/etc/resolv.conf')
  },
  toBackend: async (req: any, res: any) => {
    await axios
      .get(privateBackendUrl)
      .then((beRes) => {
        res.status(200).json({
          connectedUrl: privateBackendUrl,
          resolvedIp: beRes.request.socket.remoteAddress
        })
      })
      .catch((beErr) => {
        res.status(400).json({
          connectedUrl: privateBackendUrl,
          resolvedIp: beErr.request.socket.remoteAddress,
          errors: beErr
        })
      })
  }
}


.commitlintrc.js
module.exports = {
  rules: {
    'body-leading-blank': [1, 'always'],
    'body-max-line-length': [2, 'always', 100],
    'footer-leading-blank': [1, 'always'],
    'footer-max-line-length': [2, 'always', 100],
    'header-max-length': [2, 'always', 100],
    'subject-empty': [2, 'never'],
    'subject-full-stop': [2, 'never', '.'],
    'type-case': [2, 'always', 'lower-case'],
    'type-empty': [2, 'never'],
    'type-enum': [
      2,
      'always',
      [
        'build',
        'chore',
        'ci',
        'docs',
        'feat',
        'fix',
        'perf',
        'refactor',
        'revert',
        'style',
        'test'
      ]
    ]
  },
  prompt: {
    questions: {
      type: {
        description: "Select the type of change that you're committing",
        enum: {
          feat: {
            description: 'A new feature',
            title: 'Features',
            emoji: '✨'
          },
          fix: {
            description: 'A bug fix',
            title: 'Bug Fixes',
            emoji: '🐛'
          },
          docs: {
            description: 'Documentation only changes',
            title: 'Documentation',
            emoji: '📚'
          },
          style: {
            description:
              'Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)',
            title: 'Styles',
            emoji: '💎'
          },
          refactor: {
            description:
              'A code change that neither fixes a bug nor adds a feature',
            title: 'Code Refactoring',
            emoji: '📦'
          },
          perf: {
            description: 'A code change that improves performance',
            title: 'Performance Improvements',
            emoji: '🚀'
          },
          test: {
            description: 'Adding missing tests or correcting existing tests',
            title: 'Tests',
            emoji: '🚨'
          },
          build: {
            description:
              'Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)',
            title: 'Builds',
            emoji: '🛠'
          },
          ci: {
            description:
              'Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)',
            title: 'Continuous Integrations',
            emoji: '⚙️'
          },
          chore: {
            description: "Other changes that don't modify src or test files",
            title: 'Chores',
            emoji: '♻️'
          },
          revert: {
            description: 'Reverts a previous commit',
            title: 'Reverts',
            emoji: '🗑'
          }
        }
      },
      scope: {
        description:
          'What is the scope of this change (e.g. component or file name)'
      },
      subject: {
        description: 'Write a short, imperative tense description of the change'
      },
      body: {
        description: 'Provide a longer description of the change'
      },
      isBreaking: {
        description: 'Are there any breaking changes?'
      },
      breakingBody: {
        description:
          'A BREAKING CHANGE commit requires a body. Please enter a longer description of the commit itself'
      },
      breaking: {
        description: 'Describe the breaking changes'
      },
      isIssueAffected: {
        description: 'Does this change affect any open issues?'
      },
      issuesBody: {
        description:
          'If issues are closed, the commit requires a body. Please enter a longer description of the commit itself'
      },
      issues: {
        description: 'Add issue references (e.g. "fix #123", "re #123".)'
      }
    }
  }
}


src/index.ts
import fs from 'fs'
import path from 'path'

import express from 'express'
import favicon from 'serve-favicon'
import Helmet from 'helmet'
import morgan from 'morgan'
import swaggerStats from 'swagger-stats'
import * as OpenApiValidator from 'express-openapi-validator'

// environment variables
const instrumentationKey = process.env.APPINSIGHTS_INSTRUMENTATIONKEY
const serverPort = process.env.WEBSITES_PORT || 8080
const swaggerStatsUsername = process.env.SWAGGER_STATS_USERNAME
const swaggerStatsPassword = process.env.SWAGGER_STATS_PASSWORD

// https://docs.microsoft.com/en-us/azure/azure-monitor/app/nodejs
if (instrumentationKey) {
  const appInsights = require('applicationinsights')
  appInsights
    .setup(instrumentationKey)
    .setAutoCollectConsole(true, true)
    .setSendLiveMetrics(true)
    .start()
}

// setup Express
const app = express()
app.use(
  Helmet({
    contentSecurityPolicy: false
  })
)
app.use(favicon(path.join(__dirname, 'favicon.ico')))
app.use(express.json())

// logging - https://github.com/expressjs/morgan#options
app.use(morgan('combined'))

// monitoring - https://github.com/slanatech/swagger-stats
if (swaggerStatsUsername && swaggerStatsPassword) {
  app.use(
    swaggerStats.getMiddleware({
      swaggerSpec: fs.readFileSync(path.join(__dirname, '..', 'openapi.yaml')),
      uriPath: '/stats',
      authentication: true,
      onAuthenticate: function (_: any, username: string, password: string) {
        return (
          username === swaggerStatsUsername && password === swaggerStatsPassword
        )
      }
    })
  )
}

// routing - https://github.com/cdimascio/express-openapi-validator
app.use(
  OpenApiValidator.middleware({
    apiSpec: path.join(__dirname, '..', 'openapi.yaml'),
    operationHandlers: path.join(__dirname),
    validateRequests: {
      allowUnknownQueryParameters: false,
      removeAdditional: true
    },
    validateResponses: {
      onError: (error, body, req) => {
        console.log('Response body fails validation: ', error)
        console.log('Emitted from:', req.originalUrl)
        console.debug(body)
      }
    },
    validateFormats: 'full' // default: "fast"
  })
)

// default handler for not implemented
app.get('/*', async (req: any, res: any) => {
  await res.status(501).json({ error: 'Not implemented' })
})

// generic handler for errors
app.use(async (err: any, req: any, res: any, next: any) => {
  await res.status(err.status || 500).json({
    message: err.message,
    errors: err.errors
  })
})

// start the server
app
  .listen(serverPort, () =>
    console.info(`API listening at http://localhost:${serverPort}`)
  )
  .on('error', function (err) {
    return console.log(`Could not start API: ${err}`)
  })
